### DISCUSSION TITLE
Why is there a dot added to the path, when calling a server/api?

### DISCUSSION THREAD

**Discussion by rowild** [nuxt/nuxt#20573] [Questions]
Hi!

Learning Nuxt 3 and found a Youtube tutorial, where a `server/api/posts.js` route is created with this script:

```js
import posts from "../../data/posts.json"

export default (req, res) => {
    const { pathname } = new URL(req.url, 'http://${req.headers.host}');
    const id = pathname.split('/')[1];

    if(!id) {
       return { posts }
    }
    else {
        return posts.find(post => post.id === parseInt(id))
    }
}
```

This route is called like this: `const { posts } = await $fetch('/api/posts');`
but the console tells me this:

```sh
dex.vue:32 Uncaught (in promise) FetchError:  (404 Cannot find any route matching /api/posts. (/api/posts))
    at async setup (index.vue:32:44)
```

There seems to be a "." after `api/posts` and I have no idea where it is coming from.

I re-wrote the export statement using `defineEventHandler(event => {})` (which I believe is the more proper version) and I also renamed the `posts.js` file to `posts.get.js`, since that seems to be the way to define the method. But that does not change the problem with the dot.

Can anybody please advise me what I have to do to fix my errors? Thank you very much!

**Comment by warflash** [comment]
Pretty sure the dot in the error message is just to signal the sentence is over🤔

**Comment by Gruce** [comment]
Could you please provide a reproduction?

**Comment by maou-shonen** [comment]
I didn't check the youtube tutorial you provided,
But from the code you provided, it looks like you should want to create `~/server/api/posts/[id].get.ts`
then use `getRouterParam` or your pathname.split('/')[3] to get the ID

**Comment by rowild** [comment]
Thanks for your feedbacks!
I provided a codesandbox here, hope it works?
https://codesandbox.io/p/github/rowild/nuxt-3-test/main?workspaceId=9706af22-90e8-48a0-b279-0fe41ac2a35f

@mao-shonen I think this is not what I want, because initially I'd like to get ALL the posts. Or do I misunderstand you?

UPDATE
Hmmm... I think the repo is not accessible for outsiders... how can I change that on codesandbox? Or can you recommend another platform for code samples for Nuxt 3?

UPDATE 2:
I think this might work:
https://codesandbox.io/p/github/rowild/nuxt-3-test/

**Reply by maou-shonen** [reply]
Yes, the problem with your code is as I stated,
Your `server/api/posts.js` serves the `/api/posts` endpoint
But you request is `/api/posts/{id}`

**Reply by rowild** [reply]
Sorry, but I do not understand. In `pages/index.vue` – the startpage – I request 
`const { posts } = await $fetch('/api/posts');`

And `server/api/psots.js` is, at the moment, nothing else but
```js
export default defineEventHandler(event => {
    console.log('event =', event);
)}
```

Are you by any chance looking at `page/posts/[id].vue`? If so, I am not even there yet with my little app... :-)

**Reply by maou-shonen** [reply]
Oh...the code you provided has many pages, I didn't understand correctly,
I re-read it again, the problem with the code you provided is that /server/api/posts does not return anything,
This is not considered an endpoint in nuxt3 (I don't know why)

**Comment by rowild** [comment]
Ah! It is working! It seems the api MUST return valid data; a simple "console.log" to test it will through an error. 

But this only works, when returning `return { posts }` directly.
As soon as I start fetching the URL and the pathname etc, things get strange again. I would expect that, as long as the id is undefined, all posts get returned. But it seems `id` is always defined, so `!id` never works...

```js
export default defineEventHandler(event => {
    const { pathname } = new URL(event.url, 'http://${event.headers.host}');
    const id = pathname.split('/')[1];

    console.log('id =', id); // returns undefined
    console.log('event.headers =', event.headers); // returns undefined

    if(!id) {
       console.log("THIS SHOULD WORK, BUT IT DOES NOT... RETURN ALL POSTS")
       return { posts }
    }
    else {
       console.log("RETURN SINGLE POST ONLY")
        return posts.find(post => post.id === parseInt(id))
    }

})
```

Any idea, why?

**Reply by maou-shonen** [reply]
Because the codesandbox you just provided is a different problem...
I suggest you start by creating two separate endpoints to understand how /server works
`/server/api/posts.ts`
`/server/api/posts/[id].ts`

**Reply by rowild** [reply]
@mao-shonen Thanks for your feedback! Much appreciated! 
Not sure though, how I can get the `id` and query `posts` in `posts/[id].js`... I am having troubles finding the documentation about that. Any change you could help me out with a link?

**Reply by maou-shonen** [reply]
https://nitro.unjs.io/guide/routing#route-with-params
